src/ProxyClient/Http/HttpAdapterBase.php

```diff
@@ -0,0 +1,184 @@
+<?php
+
+namespace FOS\HttpCache\ProxyClient\Http;
+
+use FOS\HttpCache\Exception\InvalidUrlException;
+use Http\Discovery\UriFactoryDiscovery;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\UriInterface;
+
+/**
+ * Base logic for all adapters
+ *
+ * @author David Buchmann <mail@davidbu.ch>
+ */
+abstract class HttpAdapterBase implements HttpAdapterInterface
+{
+    /**
+     * Caching proxy server host names or IP addresses.
+     *
+     * @var UriInterface[]
+     */
+    private $servers;
+
+    /**
+     * Application host name and optional base URL.
+     *
+     * @var UriInterface
+     */
+    private $baseUri;
+
+    /**
+     * @param string[] $servers Caching proxy server hostnames or IP
+     *                          addresses, including port if not port 80.
+     *                          E.g. ['127.0.0.1:6081']
+     * @param string   $baseUri Default application hostname, optionally
+     *                          including base URL, for purge and refresh
+     *                          requests (optional). This is required if
+     *                          you purge and refresh paths instead of
+     *                          absolute URLs.
+     */
+    public function __construct(array $servers, $baseUri)
+    {
+        $this->setServers($servers);
+        $this->setBaseUri($baseUri);
+    }
+
+    /**
+     * Duplicate a request for each caching server
+     *
+     * @param InvalidationRequest $request The request to duplicate for each configured server
+     *
+     * @return RequestInterface[]
+     */
+    protected function multiplex(InvalidationRequest $request)
+    {
+        $requests = [];
+
+        $uri = $request->getUri();
+
+        // If a base URI is configured, try to make partial invalidation
+        // requests complete.
+        if ($this->baseUri) {
+            if ($uri->getHost()) {
+                // Absolute URI: does it already have a scheme?
+                if (!$uri->getScheme() && $this->baseUri->getScheme() !== '') {
+                    $uri = $uri->withScheme($this->baseUri->getScheme());
+                }
+            } else {
+                // Relative URI
+                if ($this->baseUri->getHost() !== '') {
+                    $uri = $uri->withHost($this->baseUri->getHost());
+                }
+
+                if ($this->baseUri->getPort()) {
+                    $uri = $uri->withPort($this->baseUri->getPort());
+                }
+
+                // Base path
+                if ($this->baseUri->getPath() !== '') {
+                    $path = $this->baseUri->getPath() . '/' . ltrim($uri->getPath(), '/');
+                    $uri = $uri->withPath($path);
+                }
+            }
+        }
+
+        // Close connections to make sure invalidation (PURGE/BAN) requests
+        // will not interfere with content (GET) requests.
+        // TODO: or is this the job of a plugin?
```

Comment:
Base URI is optional in the first place, isn't it? In this case I think you can rely on the user, since it is not required.